query-engine/core/src/interactive_transactions/transaction.rs

```diff
@@ -0,0 +1,217 @@
+#![allow(unsafe_code)]
+
+use std::pin::Pin;
+
+use super::CachedTx;
+use crate::{
+    execute_many_operations, execute_single_operation, ClosedTx, Operation, ResponseData, TransactionError, TxId,
+};
+use connector::{Connection, Transaction};
+use crosstarget_utils::time::ElapsedTimeCounter;
+use schema::QuerySchemaRef;
+use tokio::time::Duration;
+use tracing::Span;
+use tracing_futures::Instrument;
+
+#[cfg(feature = "metrics")]
+use crate::telemetry::helpers::set_span_link_from_traceparent;
+
+struct TransactionState {
+    // Note: field order is important here because fields are dropped in the declaration order.
+    // First, we drop the `cached_tx`, which may reference `conn`. Only after that we drop `conn`.
+    cached_tx: CachedTx,
+    conn: Pin<Box<dyn Connection + Send + Sync>>,
```

Comment:
Can `conn` be completely moved into `cached_tx`? Do we need to keep it alive beyond `CachedTx::Open`?

Reply:
That is actually a great point! It will also allows us to remove `try_remove_transaction` method because the connection will be automatically dropped when the transaction moves to a closed state.

Reply:
And probably `get_unchecked` + `transmute` in `start_transaction` as well?

Reply:
Yep, all done.